// coeus-router/lib/reducers/router.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

exports.default = router;

var _isEmpty = require('lodash/isEmpty');

var _isEmpty2 = _interopRequireDefault(_isEmpty);

var _map = require('lodash/map');

var _map2 = _interopRequireDefault(_map);

var _fromPairs = require('lodash/fromPairs');

var _fromPairs2 = _interopRequireDefault(_fromPairs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function searchParse(str) {
  if ((0, _isEmpty2.default)(str)) {
    return {};
  }

  str = str.substr(1);
  return (0, _fromPairs2.default)((0, _map2.default)(str.split('&'), function (e) {
    var _e$split = e.split('=');

    var _e$split2 = _slicedToArray(_e$split, 2);

    var key = _e$split2[0];
    var value = _e$split2[1];

    return [decodeURIComponent(key), decodeURIComponent(value)];
  }));
}

function searchStringify(obj) {
  if ((0, _isEmpty2.default)(obj)) {
    return '';
  }

  return '?' + (0, _map2.default)(obj, function (v, k) {
    return [encodeURIComponent(k), encodeURIComponent(v)].join('=');
  }).join('&');
}

// a little hack to sync the popstate event
function syncPopState() {
  return new Promise(function (resolve) {
    var restore = window.onpopstate;

    window.onpopstate = function () {
      window.onpopstate = restore;
      resolve();
    };
  });
}

function router(routes) {
  return function reducer(state, action) {
    if (state === undefined) {
      var _ret = function () {
        var path = window.location.pathname;
        var search = searchParse(window.location.search);

        return {
          v: routes.match(path).then(function (result) {
            if (result === false) {
              throw new Error('404 Not Found: This should be handled in server. ' + 'A land page rendered in browser should not be 404.');
            }

            var components = result.components;
            var args = result.args;

            return {
              components: components,
              args: args,
              search: search
            };
          })
        };
      }();

      if ((typeof _ret === 'undefined' ? 'undefined' : _typeof(_ret)) === "object") return _ret.v;
    }

    switch (action.type) {
      case 'ROUTE':
        {
          var _ret2 = function () {
            var path = action.path;

            return {
              v: routes.match(path).then(function (result) {
                if (result === false) {
                  var _components = state.components;
                  var _args = state.args;
                  var _search = state.search;

                  return {
                    components: _components,
                    args: _args,
                    search: _search,
                    notFound: path
                  };
                }

                var search = action.search || {};

                history.pushState(null, null, path + searchStringify(search));

                var components = result.components;
                var args = result.args;

                return {
                  components: components,
                  args: args,
                  search: search
                };
              })
            };
          }();

          if ((typeof _ret2 === 'undefined' ? 'undefined' : _typeof(_ret2)) === "object") return _ret2.v;
        }
      case 'ROUTE_BACK':
        {
          var p = syncPopState();

          history.back();

          return p.then(function () {
            return routes.match(window.location.pathname);
          }).then(function (result) {
            // All 404 will not be pushed in history
            // so it's impossible to get false result
            // here.

            var components = result.components;
            var args = result.args;

            var search = searchParse(window.location.search);
            return {
              components: components,
              args: args,
              search: search
            };
          });
        }
      case 'ROUTE_FORWARD':
        {
          var _p = syncPopState();

          history.forward();

          return _p.then(function () {
            return routes.match(window.location.pathname);
          }).then(function (result) {
            // All 404 will not be pushed in history
            // so it's impossible to get false result
            // here.

            var components = result.components;
            var args = result.args;

            var search = searchParse(window.location.search);
            return {
              components: components,
              args: args,
              search: search
            };
          });
        }
      default:
        return state;
    }
  };
}
